File: tests/Web.AcceptanceTests/Pages/ExerciseLibraryPage.cs
namespace LiftAndShift.Web.AcceptanceTests.Pages;

public class ExerciseLibraryPage(IPage page) : BasePage(page)
{
    public override string PagePath => $"{BaseUrl}/exercise-library";

    public Task AssertHeading() => Assertions.Expect(Page.Locator("h1")).ToHaveTextAsync("Exercise Library");

    public Task Search(string term) => Page.FillAsync("input[type='search']", term);

    private ILocator ExerciseNameCell(string name) =>
        Page.Locator("tbody tr td:first-child strong", new() { HasText = name });

    private ILocator ExerciseRow(string name) =>
        Page.Locator("tbody tr").Filter(new() { Has = Page.Locator("td:first-child strong", new() { HasText = name }) });

    public Task AssertExerciseRowVisible(string name) =>
        Assertions.Expect(ExerciseNameCell(name)).ToBeVisibleAsync();

    public Task AssertExerciseRowHidden(string name) =>
        Assertions.Expect(ExerciseNameCell(name)).ToHaveCountAsync(0);

    public Task OpenAddCustomExercise() => Page.Locator("summary:has-text('Add Custom Exercise')").ClickAsync();

    public Task SetCustomExerciseName(string name) => Page.GetByLabel("Name *").FillAsync(name);

    public Task ClickAddExercise() => Page.Locator("button:has-text('Add Exercise')").ClickAsync();

    public Task ClickDeleteFor(string name) =>
        ExerciseRow(name).Locator("button:has-text('Delete')").ClickAsync();
}

File: src/Web/ClientApp/e2e/pages/exercise-library.page.ts
import { expect, type Locator } from '@playwright/test';
import { BasePage, baseUrl } from './base-page';

export class ExerciseLibraryPage extends BasePage {
  get pagePath(): string {
    return `${baseUrl()}/exercise-library`;
  }

  assertHeading() {
    return expect(this.page.locator('h1')).toHaveText('Exercise Library');
  }

  search(term: string) {
    return this.page.fill("input[type='search']", term);
  }

  private exerciseNameCell(name: string): Locator {
    return this.page.locator('tbody tr td:first-child strong', { hasText: name });
  }

  private exerciseRow(name: string): Locator {
    return this.page
      .locator('tbody tr')
      .filter({ has: this.page.locator('td:first-child strong', { hasText: name }) });
  }

  assertExerciseRowVisible(name: string) {
    return expect(this.exerciseNameCell(name)).toBeVisible();
  }

  assertExerciseRowHidden(name: string) {
    return expect(this.exerciseNameCell(name)).toHaveCount(0);
  }

  openAddCustomExercise() {
    return this.page.locator("summary:has-text('Add Custom Exercise')").click();
  }

  setCustomExerciseName(name: string) {
    return this.page.getByLabel('Name *').fill(name);
  }

  clickAddExercise() {
    return this.page.locator("button:has-text('Add Exercise')").click();
  }

  clickDeleteFor(name: string) {
    return this.exerciseRow(name).locator("button:has-text('Delete')").click();
  }
}
